**Considered problem:** In this notebook, we aim to forecast the power output of a solar farm for an extended period of approximately 150 days. By generating these predictions, we seek to facilitate improved planning and decision-making processes for the future management and operation of the solar farm.

**Purpose / research question:** This notebook aims to achieve two primary objectives. The first goal is to establish a baseline model, which will serve as a reference point for comparing the performance of subsequent models. At this stage, the complexity and performance of the baseline model are not of paramount importance, as the primary focus is to create a foundation for future comparisons.

The second objective is to explore the implementation of a stacking approach. This process involves two main steps: initially, we will experiment with an extensive collection of models to assess their performance. Subsequently, we will narrow down the selection based on the results obtained, allowing us to identify the most suitable candidates for the stacking ensemble.

The notebook is structured as follows:

<a id='top'></a>
[1. Data Inspection, Clean-Up & Preprocessing](#data-inspection) <br>
&emsp; [1.1 Load PV_01 Data and first look](#first-look) <br>
&emsp; [1.2 Dataset Overview](#overview) <br>
&emsp; [1.3 Feature reduction](#feature-reduction)<br>
&emsp; [1.4 Data deepdive](#deep-dive)<br>
[2. Linear regression baseline model](#baseline) <br>
[3. Stacking](#stacking) <br>
&emsp; [3.1 Initial model](#big-model) <br>
&emsp; [3.2 Model inspection and elimination](#model-inspection) <br>
&emsp; [3.3 Final results](#final-results) <br>
[4. Comparison](#comp) <br>
[5. Conclusion](#conclusion) <br>
&emsp; [5.1 Summary](#summary) <br>
&emsp; [5.2 Further Ideas](#ideas) <br>

In [44]:
import pandas as pd
import numpy as np
import plotly.express as px
import plotly.graph_objects as go

import pickle

from sklearn.linear_model import LinearRegression
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, r2_score

import utlis
seed = 42
np.random.seed(seed)



# turn off (irrelevant) warnings
import warnings
warnings.filterwarnings(action='ignore', category=FutureWarning)


# 1. Data Inspection, Clean-Up & Preprocessing<a id='data-inspection'></a>
[to the Overview](#top)

## 1.1 Load PV_01 Data and first look <a id='first-look'></a>

In [45]:
df = pd.read_csv('GermanSolarFarm/data/pv_01.csv', delimiter=';')

In [46]:
# Lock the dataframe and display all columns
with pd.option_context('display.max_columns', None):
    display(df.head(20))

time_idx  hour_of_day  hour_of_day_cos  hour_of_day_sin  month_of_year  \
0        1.0     0.000000         1.000000         0.000000       0.909091   
1        2.0     0.142857         0.957824         0.399332       0.909091   
2        3.0     0.285714         0.838278         0.732544       0.909091   
3        4.0     0.428571         0.661157         0.944463       0.909091   
4        5.0     0.571429         0.455788         1.000000       0.909091   
5        6.0     0.714286         0.256176         0.889960       0.909091   
6        7.0     0.857143         0.095371         0.632563       0.909091   
7        8.0     1.000000         0.000000         0.270427       0.909091   
8        9.0     0.000000         1.000000         0.000000       0.909091   
9       10.0     0.142857         0.957824         0.399332       0.909091   
10      11.0     0.285714         0.838278         0.732544       0.909091   
11      12.0     0.428571         0.661157         0.944463       0.909091   
12      13.0     0.571429         0.455788         1.000000       0.909091   
13      14.0     0.714286         0.256176         0.889960       0.909091   
14      15.0     0.857143         0.095371         0.632563       0.909091   
15      16.0     1.000000         0.000000         0.270427       0.909091   
16      17.0     0.000000         1.000000         0.000000       0.909091   
17      18.0     0.142857         0.957824         0.399332       0.909091   
18      19.0     0.285714         0.838278         0.732544       0.909091   
19      20.0     0.428571         0.661157         0.944463       0.909091   

    month_of_year_cos  month_of_year_sin  season_of_year  season_of_year_cos  \
0            0.017332           0.258819             1.0                 0.0   
1            0.017332           0.258819             1.0                 0.0   
2            0.017332           0.258819             1.0                 0.0   
3            0.017332           0.258819             1.0                 0.0   
4            0.017332           0.258819             1.0                 0.0   
5            0.017332           0.258819             1.0                 0.0   
6            0.017332           0.258819             1.0                 0.0   
7            0.017332           0.258819             1.0                 0.0   
8            0.017332           0.258819             1.0                 0.0   
9            0.017332           0.258819             1.0                 0.0   
10           0.017332           0.258819             1.0                 0.0   
11           0.017332           0.258819             1.0                 0.0   
12           0.017332           0.258819             1.0                 0.0   
13           0.017332           0.258819             1.0                 0.0   
14           0.017332           0.258819             1.0                 0.0   
15           0.017332           0.258819             1.0                 0.0   
16           0.017332           0.258819             1.0                 0.0   
17           0.017332           0.258819             1.0                 0.0   
18           0.017332           0.258819             1.0                 0.0   
19           0.017332           0.258819             1.0                 0.0   

    season_of_year_sin  sunposition_thetaZ  sunposition_solarAzimuth  \
0                  0.0            1.000000                  0.250553   
1                  0.0            1.000000                  0.250817   
2                  0.0            1.000000                  0.251081   
3                  0.0            0.708461                  0.379841   
4                  0.0            0.642045                  0.555848   
5                  0.0            0.906693                  0.715489   
6                  0.0            1.000000                  0.747869   
7                  0.0            1.000000                  0.747607   
8                  0.0            1.

In [47]:
# Check unique values in the 'Unnamed: 51' column
df['Unnamed: 51'].unique()

array([nan])

After checking the data, this column seems to be an encoding issue because every row ends with a ';.' The column can, therfore, be dropped. 

In [48]:
# Drop the 'Unnamed: 51' column
df = df.drop('Unnamed: 51', axis=1)

Firsty the entire datset apart from the target are scaled from 0 to 1 using a min-max normalization. The target on the other hand was scaled using the ouput capacity of the entire solarfarm. This should in theory negate the effects of different nominal capacitys. 

Explanation for every coloumn (Unfortunatly this was not provided therefore some interpretations might be flaud):
Every values is scaled between 0 - 1, and also contain a sign and cosin coding. We will focus on the value itself once. Some values also contain a Math postfix it is unclear what is meant by that 


## 1.2 Dataset Overview <a id='overview'></a>

The dataset consists of multiple features (columns) that have been preprocessed to ensure consistency and enable effective modeling. The preprocessing steps include scaling the values and encoding specific features.

### Preprocessing

1. **Min-max normalization**: All features, except for the target variable, have been scaled from 0 to 1 using min-max normalization. This process transforms the original values in each column to a standardized scale, making it easier for machine learning algorithms to converge faster and improve overall performance.

2. **Target variable scaling**: The target variable, which seems to be related to solar farm output, has been scaled using the output capacity of the entire solar farm. This approach normalizes the target variable across different solar farms with varying nominal capacities, enabling fair comparisons and analysis.

### Overall column Explanations

1. **Sine and cosine encoding**: Some of the features in the dataset have been encoded using sine and cosine transformations. This is often done for cyclical features, such as time, to capture their cyclical nature and make it easier for the model to understand the underlying patterns.

2. **Mathematical postfix**: Certain values in the dataset have a mathematical postfix. Without additional context, it's difficult to accurately interpret the meaning of these postfixes. It could be a notation system used to differentiate between different types of values or measurements in the dataset, but further information would be required to provide a definitive interpretation.

### Column Explanation
| Column Name                      | Description                                                                                                                                                                                                                                                                                                                                                        |
|----------------------------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| time_idx                         | A unique identifier for each point in the data.                                                                                                                                                                                                                                                                                                                    |
| hour_of_day                      | The hour of the day the data was collected ranged from 0 to 1 (3-hour resolution).                                                                                                                                                                                                                                                                                 |
| month_of_year                    | The month of the year the data was collected.                                                                                                                                                                                                                                                                                                                      |
| season_of_year                   | The season of the year the data was collected.                                                                                                                                                                                                                                                                                                                     |
| sunposition_thetaZ               | The zenith angle of the sun in degrees, which is the angle between the sun and the vertical. https://pvpmc.sandia.gov/modeling-steps/1-weather-design-inputs/sun-position/                                                                                                                                                                                         |
| sunposition_solarAzimuth         | The solar azimuth angle is the azimuth (horizontal angle with respect to north) of the Sun's position. This horizontal coordinate defines the Sun's relative direction along the local horizon, whereas the solar zenith angle (or its complementary angle solar elevation) defines the Sun's apparent altitude. https://en.wikipedia.org/wiki/Solar_azimuth_angle |
| sunposition_extraTerr            | Extraterrestrial radiation () is the intensity (power) of the sun at the top of the Earth’s atmosphere. https://pvpmc.sandia.gov/modeling-steps/1-weather-design-inputs/irradiance-and-insolation-2/extraterrestrial-radiation/                                                                                                                                    |
| sunposition_solarHeight          | The height of the sun above the horizon in degrees. https://en.wikipedia.org/wiki/Position_of_the_Sun#:~:text=At%20the%20solstices%2C%20the%20angle,at%20the%20southern%20summer%20solstice.                                                                                                                                                                       |
| clearsky_diffuse                 | The amount of direct radiation on a horizontal surface under clear sky conditions in W/m^2.                                                                                                                                                                                                                                                                        |
| clearsky_direct                  | The total amount of radiation on a horizontal surface under clear sky conditions in W/m^2.                                                                                                                                                                                                                                                                         |
| clearsky_global                  | The amount of diffuse radiation on a horizontal surface under clear sky conditions in W/m^2, likely aggregated over the day.                                                                                                                                                                                                                                       |
| clearsky_diffuse_agg             | The amount of diffuse radiation on a horizontal surface under clear sky conditions in W/m^2, likely aggregated over the day.                                                                                                                                                                                                                                       |
| clearsky_direct_agg              | The amount of direct radiation on a horizontal surface under clear sky conditions in W/m^2, likely aggregated over the day.                                                                                                                                                                                                                                        |
| clearsky_global_agg              | The total amount of radiation on a horizontal surface under clear sky conditions in W/m^2, likely aggregated over the day.                                                                                                                                                                                                                                         |
| Albedo                           | The fraction of solar energy reflected by a surface, ranging from 0 to 1. https://en.wikipedia.org/wiki/Albedo                                                                                                                                                                                                                                                     |
| WindComponentUat0                | Not 100% sure but likely windspeed at the ground in a certain direction in m/s. http://colaweb.gmu.edu/dev/clim301/lectures/wind/wind-uv                                                                                                                                                                                                                           |
| WindComponentVat0                | Not 100% sure but likely windspeed at the ground in a certain direction in m/s.  http://colaweb.gmu.edu/dev/clim301/lectures/wind/wind-uv                                                                                                                                                                                                                          |
| WindComponentUat100              | Not 100% sure but likely windspeed at the ground in a certain direction in m/s.  http://colaweb.gmu.edu/dev/clim301/lectures/wind/wind-uv                                                                                                                                                                                                                          |
| WindComponentVat100              | Not 100% sure but likely windspeed at the ground in a certain direction in m/s.  http://colaweb.gmu.edu/dev/clim301/lectures/wind/wind-uv                                                                                                                                                                                                                          |
| DewpointTemperatureAt0           | The temperature at which dew would form at ground level in degrees Celsius.                                                                                                                                                                                                                                                                                        |
| TemperatureAt0                   | The temperature at ground level.                                                                                                                                                                                                                                                                                                                                   |
| PotentialVorticityAt1000         | Likely: The potential vorticity at a pressure level of 1000 hPa in PVU (potential vorticity unit). https://en.wikipedia.org/wiki/Potential_vorticity                                                                                                                                                                                                               |
| PotentialVorticityAt950          | Likely: The potential vorticity at a pressure level of 950 hPa in PVU (potential vorticity unit).  https://en.wikipedia.org/wiki/Potential_vorticity                                                                                                                                                                                                               |
| RelativeHumidityAt1000           | Likely: The relative humidity at a pressure level of 1000 hPa. https://www.lenntech.com/calculators/humidity/relative-humidity.htm                                                                                                                                                                                                                                 |
| RelativeHumidityAt950            | Likely: The relative humidity at a pressure level of 950 hPa, expressed as a percentage. https://www.lenntech.com/calculators/humidity/relative-humidity.htm                                                                                                                                                                                                       |
| RelativeHumidityAt0              | The relative humidity at ground level.                                                                                                                                                                                                                                                                                                                             |
| SnowDensityAt0                   | The density of snow on the ground.                                                                                                                                                                                                                                                                                                                                 |
| SnowDepthAt0                     | The depth of snow on the ground.                                                                                                                                                                                                                                                                                                                                   |
| SnowfallPlusStratiformSurfaceAt0 | The amount of snowfall and stratiform precipitation at ground level.                                                                                                                                                                                                                                                                                               |
| SurfacePressureAt0               | The air pressure at ground level.                                                                                                                                                                                                                                                                                                                                  |
| SolarRadiationGlobalAt0          | The total amount of solar radiation on a horizontal surface at ground level.                                                                                                                                                                                                                                            |
| SolarRadiationDirectAt0          | The amount of diffuse radiation on a horizontal surface at ground level.                                                                                                                                                                                                                                                                                  |
| SolarRadiationDiffuseAt0         | The amount of diffuse radiation on a horizontal surface at ground.                                                                                                                                                                                                                                                                                                 |
| TotalCloudCoverAt0               | The total cloud cover at ground level.                                                                                                                                                                                                                                                                                                                             |
| LowerWindSpeed                   | The average wind speed in (likely) the lower atmosphere.                                                                                                                                                                                                                                                                                                           |
| LowerWindDirection               | The average wind direction in the lower atmosphere.                                                                                                                                                                                                                                                                                                                |
| LowerWindDirectionMath           | The mathematical wind direction in the lower atmosphere.                                                                                                                                                                                                                                                                                                           |
| UpperWindSpeed                   | The average wind speed in (likely) the upper atmosphere.                                                                                                                                                                                                                                                                                                           |
| UpperWindDirection               | The average wind direction in the upper atmosphere.                                                                                                                                                                                                                                                                                                                |
| UpperWindDirectionMath           | Likely the mathematical wind direction in the upper atmosphere in degrees.                                                                                                                                                                                                                                                                                         |
| power_normed                     | The normalized power output of the solar panels, ranging from 0 to 1.                                                                                                                                                                                                                                                                                        


In summary, the dataset has undergone various preprocessing steps to ensure consistency and facilitate effective modeling. The features have been scaled and encoded as necessary, but some column descriptions remain unclear due to the lack of information.





### Size of the dataset


In [49]:
# Get the length of the dataframe
print(f"Total number of measurements: {len(df)}")
# Devide the dataframe length by 8 to get the number of days (8 measurements per day = a resolution of 3 hours)
print(f"Number of days measured: {len(df)/8}")

Total number of measurements: 6217
Number of days measured: 777.125


This is a little odd as the lecture (2-Energy p. 41) stated 990 days however the data provided only seems to include 777 days. There does not seem to be a predefined train test split. 

### Datatypes in the dataset

In [50]:
# Lets check the data types of all the columns
# dont restict the output rows
with pd.option_context('display.max_rows', None):
    display(df.dtypes)

time_idx                            float64
hour_of_day                         float64
hour_of_day_cos                     float64
hour_of_day_sin                     float64
month_of_year                       float64
month_of_year_cos                   float64
month_of_year_sin                   float64
season_of_year                      float64
season_of_year_cos                  float64
season_of_year_sin                  float64
sunposition_thetaZ                  float64
sunposition_solarAzimuth            float64
sunposition_extraTerr               float64
sunposition_solarHeight             float64
clearsky_diffuse                    float64
clearsky_direct                     float64
clearsky_global                     float64
clearsky_diffuse_agg                float64
clearsky_direct_agg                 float64
clearsky_global_agg                 float64
Albedo                              float64
WindComponentUat0                   float64
WindComponentVat0               

The output looks good as we would expect all coloums to have a numeric value which is the case.

In [51]:
df.head()

time_idx  hour_of_day  hour_of_day_cos  hour_of_day_sin  month_of_year  \
0       1.0     0.000000         1.000000         0.000000       0.909091   
1       2.0     0.142857         0.957824         0.399332       0.909091   
2       3.0     0.285714         0.838278         0.732544       0.909091   
3       4.0     0.428571         0.661157         0.944463       0.909091   
4       5.0     0.571429         0.455788         1.000000       0.909091   

   month_of_year_cos  month_of_year_sin  season_of_year  season_of_year_cos  \
0           0.017332           0.258819             1.0                 0.0   
1           0.017332           0.258819             1.0                 0.0   
2           0.017332           0.258819             1.0                 0.0   
3           0.017332           0.258819             1.0                 0.0   
4           0.017332           0.258819             1.0                 0.0   

   season_of_year_sin  ...  LowerWindDirection  LowerWindDirectionMath  \
0                 0.0  ...            0.558633                0.190752   
1                 0.0  ...            0.531501                0.217852   
2                 0.0  ...            0.623444                0.126020   
3                 0.0  ...            0.669254                0.080266   
4                 0.0  ...            0.698923                0.050633   

   LowerWindDirectionCos  LowerWindDirectionSin  UpperWindSpeed  \
0               0.681740               0.965801        0.359087   
1               0.600192               0.989859        0.475393   
2               0.851167               0.855924        0.472248   
3               0.937664               0.741765        0.591015   
4               0.974850               0.656581        0.641152   

   UpperWindDirection  UpperWindDirectionMath  UpperWindDirectionCos  \
0            0.577455                0.172559               0.733785   
1            0.545543                0.204424               0.641217   
2            0.635529                0.114573               0.875928   
3            0.673493                0.076666               0.943093   
4            0.701128                0.049073               0.976413   

   UpperWindDirectionSin  power_normed  
0               0.941978      0.000000  
1               0.979643      0.000000  
2               0.829663      0.000000  
3               0.731664      0.035185  
4               0.651756      0.120988  

[5 rows x 51 columns]

## 1.3 Feature reduction <a id='feature-reduction'></a>
Given the large number of features in the dataset, some of which may be difficult to interpret without domain knowledge in meteorology, it can be challenging to identify meaningful relationships among the variables. Therefore, in order to facilitate model interpretability and potentially improve prediction performance, we propose reducing the number of features. Specifically, we will focus on selecting features that have been previously identified as having a significant impact on model performance, as discussed in relevant lecture. This process will allow us to mitigate the "curse of dimensionality" and improve the efficiency of our modeling efforts.


We propose excluding certain columns from the dataset for the following reasons:

- <b>"WindComponentUat0, WindComponentVat0, WindComponentUat100, and WindComponentVat100"</b> will be excluded as their meaning and impact on power output prediction are unclear. Instead, we will rely on the more easily interpretable LowerWindSpeed and UpperWindSpeed columns for information about wind speed.

- <b>"clearsky_diffuse, clearsky_direct, clearsky_global, clearsky_diffuse_agg, clearsky_direct_agg, and clearsky_global_agg"</b> will be excluded due to their potential difficulty in interpretation and dependence on cloud cover. Cloud cover, which is measured separately in the TotalCloudCoverAt0 column, can be a more informative metric in predicting solar power output.

- <b>"RelativeHumidityAt1000 and RelativeHumidityAt950"</b> columns, while containing potentially valuable information, may be difficult for the model to interpret due to the use of three different pressure measurements at varying altitudes. Therefore, we will focus solely on the RelativeHumidityAt0 column, which is measured at ground level.

- <b>"PotentialVorticityAt950"</b> will be excluded since the PotentialVorticityAt1000 measurement should be sufficient for modeling purposes.



In [52]:
# drop indComponentUat0, WindComponentVat0, WindComponentUat100, and WindComponentVat100
df = df.drop(['WindComponentUat0', 'WindComponentVat0', 'WindComponentUat100', 'WindComponentVat100'], axis=1)
# drop clearsky_diffuse, clearsky_direct, clearsky_global, clearsky_diffuse_agg, clearsky_direct_agg, and clearsky_global_agg
df = df.drop(['clearsky_diffuse', 'clearsky_direct', 'clearsky_global', 'clearsky_diffuse_agg', 'clearsky_direct_agg', 'clearsky_global_agg'], axis=1)
# drop RelativeHumidityAt1000 and RelativeHumidityAt950
df = df.drop(['RelativeHumidityAt1000', 'RelativeHumidityAt950'], axis=1)
# drop PotentialVorticityAt950
df = df.drop(['PotentialVorticityAt950'], axis=1)

In [53]:
# Lock the dataframe and display all columns
with pd.option_context('display.max_columns', None):
    display(df.head(20))

time_idx  hour_of_day  hour_of_day_cos  hour_of_day_sin  month_of_year  \
0        1.0     0.000000         1.000000         0.000000       0.909091   
1        2.0     0.142857         0.957824         0.399332       0.909091   
2        3.0     0.285714         0.838278         0.732544       0.909091   
3        4.0     0.428571         0.661157         0.944463       0.909091   
4        5.0     0.571429         0.455788         1.000000       0.909091   
5        6.0     0.714286         0.256176         0.889960       0.909091   
6        7.0     0.857143         0.095371         0.632563       0.909091   
7        8.0     1.000000         0.000000         0.270427       0.909091   
8        9.0     0.000000         1.000000         0.000000       0.909091   
9       10.0     0.142857         0.957824         0.399332       0.909091   
10      11.0     0.285714         0.838278         0.732544       0.909091   
11      12.0     0.428571         0.661157         0.944463       0.909091   
12      13.0     0.571429         0.455788         1.000000       0.909091   
13      14.0     0.714286         0.256176         0.889960       0.909091   
14      15.0     0.857143         0.095371         0.632563       0.909091   
15      16.0     1.000000         0.000000         0.270427       0.909091   
16      17.0     0.000000         1.000000         0.000000       0.909091   
17      18.0     0.142857         0.957824         0.399332       0.909091   
18      19.0     0.285714         0.838278         0.732544       0.909091   
19      20.0     0.428571         0.661157         0.944463       0.909091   

    month_of_year_cos  month_of_year_sin  season_of_year  season_of_year_cos  \
0            0.017332           0.258819             1.0                 0.0   
1            0.017332           0.258819             1.0                 0.0   
2            0.017332           0.258819             1.0                 0.0   
3            0.017332           0.258819             1.0                 0.0   
4            0.017332           0.258819             1.0                 0.0   
5            0.017332           0.258819             1.0                 0.0   
6            0.017332           0.258819             1.0                 0.0   
7            0.017332           0.258819             1.0                 0.0   
8            0.017332           0.258819             1.0                 0.0   
9            0.017332           0.258819             1.0                 0.0   
10           0.017332           0.258819             1.0                 0.0   
11           0.017332           0.258819             1.0                 0.0   
12           0.017332           0.258819             1.0                 0.0   
13           0.017332           0.258819             1.0                 0.0   
14           0.017332           0.258819             1.0                 0.0   
15           0.017332           0.258819             1.0                 0.0   
16           0.017332           0.258819             1.0                 0.0   
17           0.017332           0.258819             1.0                 0.0   
18           0.017332           0.258819             1.0                 0.0   
19           0.017332           0.258819             1.0                 0.0   

    season_of_year_sin  sunposition_thetaZ  sunposition_solarAzimuth  \
0                  0.0            1.000000                  0.250553   
1                  0.0            1.000000                  0.250817   
2                  0.0            1.000000                  0.251081   
3                  0.0            0.708461                  0.379841   
4                  0.0            0.642045                  0.555848   
5                  0.0            0.906693                  0.715489   
6                  0.0            1.000000                  0.747869   
7                  0.0            1.000000                  0.747607   
8                  0.0            1.

### Check for outlieres and Missing data

In [54]:
with pd.option_context('display.max_columns', None):
    display(df.describe())

time_idx  hour_of_day  hour_of_day_cos  hour_of_day_sin  \
count  6217.000000  6217.000000      6217.000000      6217.000000   
mean   3109.000000     0.499920         0.533149         0.608563   
std    1794.837644     0.327388         0.365148         0.333564   
min       1.000000     0.000000         0.000000         0.000000   
25%    1555.000000     0.142857         0.256176         0.270427   
50%    3109.000000     0.428571         0.661157         0.632563   
75%    4663.000000     0.714286         0.957824         0.889960   
max    6217.000000     1.000000         1.000000         1.000000   

       month_of_year  month_of_year_cos  month_of_year_sin  season_of_year  \
count    6217.000000        6217.000000        6217.000000     6217.000000   
mean        0.544548           0.420717           0.600220        0.548603   
std         0.320168           0.358962           0.334571        0.378155   
min         0.000000           0.000000           0.000000        0.000000   
25%         0.272727           0.068148           0.258819        0.333333   
50%         0.545455           0.377014           0.707107        0.666667   
75%         0.818182           0.762999           0.866025        1.000000   
max         1.000000           1.000000           1.000000        1.000000   

       season_of_year_cos  season_of_year_sin  sunposition_thetaZ  \
count         6217.000000         6217.000000         6217.000000   
mean             0.394155            0.554595            0.800269   
std              0.385151            0.391409            0.279257   
min              0.000000            0.000000            0.000000   
25%              0.000000            0.000000            0.656749   
50%              0.171573            0.707107            1.000000   
75%              0.585786            0.707107            1.000000   
max              1.000000            1.000000            1.000000   

       sunposition_solarAzimuth  sunposition_extraTerr  \
count               6217.000000           6.217000e+03   
mean                   0.476646           2.287822e-01   
std                    0.287884           3.065492e-01   
min                    0.000000           0.000000e+00   
25%                    0.242300           3.391465e-18   
50%                    0.383964           4.996425e-18   
75%                    0.730422           4.209835e-01   
max                    1.000000           1.000000e+00   

       sunposition_solarHeight       Albedo  DewpointTemperatureAt0  \
count              6217.000000  6217.000000             6217.000000   
mean                  0.199731     0.280741                0.582749   
std                   0.279257     0.172684                0.151626   
min                   0.000000     0.000000                0.000000   
25%                   0.000000     0.133041                0.471233   
50%                   0.000000     0.283233                0.583336   
75%                   0.343251     0.445688                0.692286   
max                   1.000000     1.000000                1.000000   

       TemperatureAt0  PotentialVorticityAt1000  RelativeHumidityAt0  \
count     6217.000000               6217.000000          6217.000000   
mean         0.420889                  0.479249             0.751955   
std          0.160634                  0.052128             0.177530   
min          0.000000                  0.000000             0.000000   
25%          0.293500                  0.454071             0.648657   
50%          0.414431                  0.468946             0.794304   
75%          0.536525                  0.497157             0.894825   
max          1.000000                  1.000000             1.000000   

       SnowDensityAt0  SnowDepthAt0  SnowfallPlusStratiformSurfaceAt0  \
count    6.217000e+03   6217.000000                      6.217000e+03   
mean     5.842311e-02      0.015384                      1.036643e-02   
std      1.486680e-01      0.075

All looks good but to make it easier for us to explore the data some more we drop all math, sind, cos columns as they do not help us to explore the data as a human


In [55]:
# drop all columns that include cos, sin and math 
df_reduced = df.drop([col for col in df.columns if 'cos' in col.lower() or 'sin' in col.lower() or 'math' in col.lower()], axis=1)

In [56]:
df_reduced.head(2000)

time_idx  hour_of_day  month_of_year  season_of_year  \
0          1.0     0.000000       0.909091        1.000000   
1          2.0     0.142857       0.909091        1.000000   
2          3.0     0.285714       0.909091        1.000000   
3          4.0     0.428571       0.909091        1.000000   
4          5.0     0.571429       0.909091        1.000000   
...        ...          ...            ...             ...   
1995    1996.0     0.428571       0.545455        0.666667   
1996    1997.0     0.571429       0.545455        0.666667   
1997    1998.0     0.714286       0.545455        0.666667   
1998    1999.0     0.857143       0.545455        0.666667   
1999    2000.0     1.000000       0.545455        0.666667   

      sunposition_thetaZ  sunposition_solarAzimuth  sunposition_extraTerr  \
0               1.000000                  0.250553           3.654783e-18   
1               1.000000                  0.250817           3.659645e-18   
2               1.000000                  0.251081           3.664501e-18   
3               0.708461                  0.379841           3.652606e-01   
4               0.642045                  0.555848           4.449296e-01   
...                  ...                       ...                    ...   
1995            0.184403                  0.282726           8.705038e-01   
1996            0.016498                  0.560383           9.894070e-01   
1997            0.344918                  0.786007           7.315542e-01   
1998            0.793622                  0.922593           2.477557e-01   
1999            1.000000                  0.990914           1.336689e-20   

      sunposition_solarHeight    Albedo  DewpointTemperatureAt0  ...  \
0                    0.000000  0.167936                0.654981  ...   
1                    0.000000  0.167554                0.661824  ...   
2                    0.000000  0.167554                0.597057  ...   
3                    0.291539  0.167554                0.598159  ...   
4                    0.357955  0.167554                0.559800  ...   
...                       ...       ...                     ...  ...   
1995                 0.815597  0.491479                0.891961  ...   
1996                 0.983502  0.491479                0.884334  ...   
1997                 0.655082  0.491479                0.812612  ...   
1998                 0.206378  0.491479                0.847085  ...   
1999                 0.000000  0.491479                0.847109  ...   

      SurfacePressureAt0  SolarRadiationGlobalAt0  SolarRadiationDirectAt0  \
0               0.314882                 0.000000                 0.000000   
1               0.231609                 0.000000                 0.000000   
2               0.231200                 0.000000                 0.000000   
3               0.248368                 0.057197                 0.005365   
4               0.261169                 0.244397                 0.149812   
...                  ...                      ...                      ...   
1995            0.406723                 0.412532                 0.220896   
1996            0.401599                 0.719701                 0.548149   
1997            0.399199                 0.707994                 0.609710   
1998            0.426075                 0.298584                 0.218214   
1999            0.440208                 0.020961                 0.006438   

      SolarRadiationDiffuseAt0  TotalCloudCoverAt0  LowerWindSpeed  \
0                     0.000497            0.968750        0.294080   
1                     0.000497            1.000000        0.399157   
2                     0.000497            1.000000        0.410575   
3                     0.117925            0.976562        0.581901   
4                     0.267378            0.703125        0.653922   
...                        ...                 ...             ...   
1995                  0.510179            0.

## 1.4 Data deepdive <a id='deep-dive'></a>


Let's first examine the power generation over the different seasons of a year. This is a good indicator to gage how seansonal our data acctualy is. As the solar fram data is from Germany we would expect there to be a big difference between for example winter ans summer.

In [57]:
# plot the power_normed over time, overlay the season
# change the values of the season_of_year column to be 1.0 for winter, 0.33333 for spring, 0.66666 for summer, and 0.99999 for fall
df_name = df_reduced.copy()
df_name['season_of_year_name'] = df_reduced['season_of_year'].apply(lambda x: "winter" if x == 1.0 else "spring" if x == 0.0 else "summer" if round(x, 2) == 0.33 else "fall" )

fig = px.line(df_name, x='time_idx', y='power_normed', color='season_of_year_name')
fig.update_layout(
    xaxis_title="Time",
    yaxis_title="Normalized Power",
    legend_title="Season of Year",
    title="Normalized Power of a Solar Farm over Time",
    font=dict(
        family="Montserrat, monospace",
        size=12,
        color="#7f7f7f"
    )
)
fig.show()

The chart looks like we would expect with peak production taking place in summer and winter being the weakest over all. We should account for the high seasonality of our data in a later step.

Another sanity check we can perform is to look at the snow column as we would snow to not be present in summer.

In [58]:
#Plot the SnowDepthAt0 over time, overlay the season
fig = px.line(df_name, x='time_idx', y='SnowDepthAt0', color='season_of_year_name')
fig.update_layout(
    xaxis_title="Time",
    yaxis_title="Normalized Snow Depth",
    legend_title="Season of Year",
    title="Normalized Snow depth of a Solar Farm over Time",
    font=dict(
        family="Montserrat, monospace",
        size=12,
        color="#7f7f7f"
    )
)
fig.show()

This is a little odd as there seems to be snow in summer. We can double check if there was snowfall in Germany at that time. TODO: check weather for this time period 

To explore the dataset some more it will be helpful to get a better understanding of the different correlations in our data. Therfore we will calculate the correlation of every value with the target.

In [59]:
# Look at correlations between the columns and the power_normed column
df_reduced.corr()['power_normed'].sort_values(ascending=False)

power_normed                        1.000000
SolarRadiationGlobalAt0             0.928351
SolarRadiationDirectAt0             0.892979
sunposition_solarHeight             0.807046
sunposition_extraTerr               0.806146
SolarRadiationDiffuseAt0            0.800645
TemperatureAt0                      0.510028
Albedo                              0.304486
sunposition_solarAzimuth            0.232058
DewpointTemperatureAt0              0.198577
hour_of_day                         0.193858
SurfacePressureAt0                  0.064476
time_idx                            0.033198
LowerWindSpeed                      0.009979
LowerWindDirection                  0.006156
UpperWindDirection                 -0.006571
SnowfallPlusStratiformSurfaceAt0   -0.061711
SnowDepthAt0                       -0.074284
season_of_year                     -0.103832
month_of_year                      -0.108827
SnowDensityAt0                     -0.112075
TotalCloudCoverAt0                 -0.139029
UpperWindS

**Interpretation of the correlations:**<br>
- Albedo (0.304486): Albedo has a weak positive correlation with power generation, indicating that a higher surface reflectivity might result in slightly increased power generation due to increased reflected radiation.


- sunposition_solarAzimuth (0.232058): Solar azimuth has a weaker positive correlation with power generation, suggesting that the sun's position might have a minor impact on power generation, depending on the solar panel's orientation. From this, we can derive that these are likely stationary solar panels. If they could orientate themselves, this should not be a relevant factor.


- DewpointTemperatureAt0 (0.198577): There is a weak positive correlation between dew point temperature and power generation, indicating that atmospheric conditions affecting dew point temperature might have a minor influence on power generation.


- hour_of_day (0.193858): Hour of the day has a weak positive correlation with power generation. This factor is likely not as pronounced as initially expected because of the different seasons. For example, while a solar panel might produce a lot of energy at 6 pm during the summer, it may not produce any during the winter.


- SurfacePressureAt0 (0.064476): Surface pressure shows a very weak positive correlation with power generation, indicating that it might not have a significant impact on solar energy production. This is interesting as I thought the pressure could correlate with cloud cover.


- time_idx (0.033198): The time index feature has a very weak positive correlation with power generation. This is a good sign, as we would not expect this to matter due to the balance in seasons.


- UpperWindDirection (-0.006571) and LowerWindDirection (0.006156): Both upper and lower wind directions show very weak correlations with power generation, indicating that wind direction might not play a significant role in solar energy generation. It might help the model to exclude this variable.


- SnowfallPlusStratiformSurfaceAt0 (-0.061711), SnowDepthAt0 (-0.074284), and SnowDensityAt0 (-0.112075): These snow-related features show weak negative correlations with power generation, suggesting that snow accumulation might have a minor adverse impact on solar energy production. TLDR-covered panels produce less snow.


- season_of_year (-0.103832) and month_of_year (-0.108827): Both season and month of the year show weak negative correlations with power generation, which could be due to seasonal variations in solar radiation and weather conditions. I would have expected these factors to be more prominent.


- TotalCloudCoverAt0 (-0.139029): Total cloud cover has a weak negative correlation with power generation, indicating that increased cloud cover reduces solar energy production. I expected this to be a more significant factor.


- PotentialVorticityAt1000 (-0.350961): Potential vorticity at 1000 hPa has a moderate negative correlation with power generation, implying that specific atmospheric circulation patterns could negatively influence solar energy production. However, one would need more domain knowledge for a detailed explanation.


- RelativeHumidityAt0 (-0.727691): Relative humidity has a strong negative correlation with power generation, indicating that higher humidity levels might significantly reduce solar energy production. This is a little puzzling to me unless humidity indicates rain which it does not, but it could be an indicator of fog which likely would have a drastic impact on power generation. https://www.weather.gov/lmk/humidity#:~:text=If%20the%20relative%20humidity%20is,temperature%20the%20air%20is%20at.


- sunposition_thetaZ (-0.807046): The solar zenith angle has a strong negative correlation with power generation, suggesting that as the angle between the sun and the zenith increases, power generation decreases. This is likely due to less direct sunlight reaching the solar panels.


The other factors were not explained because it seems obvious that solar radiation would increase power generation.



In [60]:
# Look at correlations between the columns
corr_matrix = df_reduced.corr()

fig = px.imshow(corr_matrix, x=corr_matrix.columns, y=corr_matrix.columns,
                color_continuous_scale='RdBu_r', zmin=-1, zmax=1)
fig.show()

# 2. Linear regression baseline model <a id='baseline'></a>

In [61]:
def eval(model, X_test: pd.DataFrame, y_test: pd.DataFrame)-> dict:
    # Predict the target variable for the test data
    y_pred = model.predict(X_test)

    # Clip the predicted values to a minimum of 0
    y_pred_clipped = np.clip(y_pred, a_min=0, a_max=None)
    
    # Calculate evaluation metrics
    mse = mean_squared_error(y_test, y_pred_clipped)
    rmse = np.sqrt(mse)
    r2 = r2_score(y_test, y_pred)

    return {
        "mse": mse,
        "rmse": rmse,
        "r2": r2
    }

In [62]:
# Get data
X_train, y_train, X_test, y_test = utlis.split_data(df_reduced=df_reduced, split_rate=0.8)

# Initialize the Linear Regression model
model = LinearRegression()

# Fit the model to the training data
model.fit(X_train, y_train)

results = eval(model, X_test, y_test)

print(f"Mean Squared Error: {results.get('mse'):.4f}")
print(f"Root Mean Squared Error: {results.get('rmse'):.4f}")
print(f"R-squared: {results.get('r2'):.4f}")

Mean Squared Error: 0.0037
Root Mean Squared Error: 0.0611
R-squared: 0.8628


- **Mean Squared Error (MSE) - 0.0037:** The MSE measures the average squared difference between the actual and predicted target values. A lower value indicates better performance, with 0 being a perfect prediction. In this case, the MSE of 0.0038 suggests that the model has relatively small errors in its predictions, indicating a reasonably good fit.

- **Root Mean Squared Error (RMSE) - 0.0611:** The RMSE represents the square root of the MSE and is expressed in the same units as the target variable. It provides an estimate of the average deviation of the model's predictions from the actual values. An RMSE of 0.0615 implies that, on average, the model's predictions deviate from the actual target values by approximately 6.15% (assuming the target variable is expressed as a ratio or percentage). This value indicates a moderately accurate prediction performance.

- **R-squared - 0.8628:** R-squared is a measure of how well the model's predictions fit the actual data, ranging from 0 to 1. A higher value is better, with 1 indicating that the model perfectly predicts the target variable. An R-squared of 0.8628 means that approximately 86.28% of the variance in the target variable can be explained by the model. This relatively high R-squared value suggests that the model has a good fit to the data.

**Note:** The predictions were clipped below 0 as values below 0 dont make a lot of sense. 



In [63]:
# Lets compare the results of the Linear Regression model to the acctual values
# Get the predictions for the test data
y_pred = model.predict(X_test)

# Clip the predicted values to a minimum of 0
y_pred_clipped = np.clip(y_pred, a_min=0, a_max=None)

# Create a dataframe with the actual and predicted values
df_pred = pd.DataFrame({'Actual': y_test.values.flatten(), 'Predicted': y_pred_clipped.flatten(), "Actual - Predicted": y_test.values.flatten() - y_pred_clipped.flatten()})

# Plot the actual and predicted values
fig = px.line(df_pred, x=df_pred.index, y=['Actual', 'Predicted', 'Actual - Predicted'])
fig.update_layout(
    xaxis_title="Time",
    yaxis_title="Normalized Power",
    legend_title="Actual vs Predicted",
    title="Normalized Power of a Solar Farm over Time",
    font=dict(
        family="Montserrat, monospace",
        size=12,
        color="#7f7f7f"
    )
)
fig.show()

The initial model performance shows promising results as a starting point. Although there is no distinct weakness such as a general tendency to overpredict or underpredict the actual results.

In [64]:
# Lets look at the weights of the model and see which features are most important
# Get the weights of the model
weights = model.coef_

# Get the feature names
feature_names = X_train.columns

# Create a dataframe with the weights and feature names
df_weights = pd.DataFrame({'Feature': feature_names, 'Weight': weights})

# Sort the dataframe by the weights
df_weights = df_weights.sort_values(by='Weight', ascending=False)

# Plot the weights
fig = px.bar(df_weights, x='Weight', y='Feature', orientation='h')
fig.update_layout(
    xaxis_title="Weight",
    yaxis_title="Feature",
    legend_title="",
    title="Feature Weights",
    font=dict(
        family="Montserrat, monospace",
        size=12,
        color="#7f7f7f"
    )
)
fig.show()


The model's assigned weights to individual features appear reasonable; however, it is noteworthy that the "hour of the day" feature seems to have minimal impact on the model's predictions. While the correlation analysis conducted earlier revealed a modest correlation for this feature, the current model does not appear to capitalize on its potential influence.

Additionally, it is intriguing to observe the significance attributed to the "dewpoint temperature" feature by the model, despite having a nearly identical correlation coefficient to that of the "hour of the day" feature. 

An analysis of its generalization performance is needed to get a better idea of the model's performance. For this purpose, it should evaluate every Dataset.

In [65]:
def eval_multiple_files(model):
    results = []
    for i in range(21):
        # open csv file
        filename = f"GermanSolarFarm/data/pv_{str((i+1)).zfill(2)}.csv"
        df = pd.read_csv(filename, delimiter=';')
        df_reduced = utlis.reduce_dataframe(df=df)
        X_train, y_train, X_test, y_test = utlis.split_data(df_reduced=df_reduced, split_rate=0.8)

        results.append({
            "file": f"pv_{i+1}",
            "results": eval(model, X_test, y_test)
            })
    return results


In [66]:
results_linear = eval_multiple_files(model)
# calulate averages
avg_mse, avg_rmse, avg_r2 = utlis.calculate_average_performance(results_linear)

print(f"Average mean Squared Error: {avg_mse:.4f}")
print(f"Average root Mean Squared Error: {avg_rmse:.4f}")
print(f"Average R-squared: {avg_r2:.4f}")


Average mean Squared Error: 0.0063
Average root Mean Squared Error: 0.0746
Average R-squared: 0.8080


In [67]:
# Extract the filenames and the results into separate lists
filenames = [item["file"] for item in results_linear]
mre_values = [item["results"]["mse"] for item in results_linear]
r2_values = [item["results"]["r2"] for item in results_linear]
rmse_values = [item["results"]["rmse"] for item in results_linear]

# Create a grouped bar chart using Plotly
fig = go.Figure(data=[
    go.Bar(name='MSE', x=filenames, y=mre_values),
    go.Bar(name='RMSE', x=filenames, y=rmse_values)
])

# Customize the layout
fig.update_layout(
    barmode='group',
    title='Results for Different Files',
    xaxis_title='Filename',
    yaxis_title='Value',
    legend_title='Metrics',
)

# Show the plot
fig.show()

We anticipated observing variations between the different files; however, the emergence of up to 3.5 times variation was unexpected. Such significant discrepancies warrant further investigation to comprehend their underlying causes and potential impacts on the model's performance. Although exploring this matter is beyond the scope of the current work, it should be considered a priority for future research to ensure a comprehensive understanding of the model's behavior and to identify potential areas for improvement.

In [68]:
# Create a grouped bar chart using Plotly
fig = go.Figure(data=[
    go.Bar(name='R2', x=filenames, y=r2_values),
])

# Customize the layout
fig.update_layout(
    barmode='group',
    title='R2 results for Different Files',
    xaxis_title='Filename',
    yaxis_title='Value',
    legend_title='Metrics',
)

# Show the plot
fig.show()

## Bagging approach 

from sklearn.ensemble import BaggingRegressor
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.model_selection import RandomizedSearchCV
from sklearn.preprocessing import StandardScaler
import utlis
import numpy as np

# Get data
X_train, y_train, X_test, y_test = utlis.split_data(df_reduced=df_reduced, split_rate=0.8)

# Define the base models to use in the ensemble
models = [
    LinearRegression(),
    Ridge(random_state=seed),
    Lasso(random_state=seed),
]

# Define the parameter grid for the Bagging Regressor
param_grid = {
    'n_estimators': np.arange(5, 50, 5),
    'max_samples': np.arange(0.5, 1.0, 0.1),
    'base_estimator': models,
    'base_estimator__fit_intercept': [True, False]
}

# Initialize the Bagging Regressor
bagging_model = BaggingRegressor(random_state=seed)

# Standardize the input features
scaler = StandardScaler()
X_train_std = scaler.fit_transform(X_train)
X_test_std = scaler.transform(X_test)

# Initialize the Randomized Search Cross-Validator with 10-fold CV and 20 iterations
random_search = RandomizedSearchCV(bagging_model, param_distributions=param_grid, n_iter=20, cv=10, random_state=seed)

# Fit the Randomized Search Cross-Validator to the training data
random_search.fit(X_train_std, y_train)

# Get the best Bagging Regressor model and its hyperparameters
best_model = random_search.best_estimator_
best_params = random_search.best_params_

# Make predictions on the test set using the best model
y_pred = best_model.predict(X_test_std)

# Evaluate the model performance
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

print(f"Best parameters: {best_params}")
print(f"Mean Squared Error: {mse:.4f}")
print(f"Root Mean Squared Error: {rmse:.4f}")
print(f"R-squared: {r2:.4f}")


# 3. Stacking <a id='stacking'></a>
## 3.1 Initial model <a id='big-model'></a>

In [ ]:
from sklearn.ensemble import StackingRegressor
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, r2_score
import utlis

# Get data
X_train, y_train, X_test, y_test = utlis.split_data(df_reduced=df_reduced, split_rate=0.8)

# Define the base models to use in the ensemble
base_models = [
    ('lr', LinearRegression()),
    ('ridge', Ridge(random_state=seed)),
    ('lasso', Lasso(random_state=seed)),
    ('dt', DecisionTreeRegressor(random_state=seed)),
    ('rf', RandomForestRegressor(random_state=seed)),
    ('svr', SVR(kernel='linear'))
]

# Define the meta-model to use in the ensemble
meta_model = RandomForestRegressor()

# Initialize the Stacking Regressor
stacking_model = StackingRegressor(estimators=base_models, final_estimator=meta_model)

# Fit the Stacking Regressor to the training data
stacking_model.fit(X_train, y_train)

# Make predictions on the test set using the Stacking Regressor
y_pred = stacking_model.predict(X_test)

# Evaluate the performance of the Stacking Regressor using mean squared error, root mean squared error, and R-squared
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse:.4f}")
print(f"Root Mean Squared Error: {rmse:.4f}")
print(f"R-squared: {r2:.4f}")


Mean Squared Error: 0.0038
Root Mean Squared Error: 0.0614
R-squared: 0.8631


In [71]:
from sklearn.ensemble import StackingRegressor
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor
from sklearn.svm import SVR
from sklearn.metrics import mean_squared_error, r2_score
import utlis
from sklearn.neural_network import MLPRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.gaussian_process import GaussianProcessRegressor

# Get data
X_train, y_train, X_test, y_test = utlis.split_data(df_reduced=df_reduced, split_rate=0.8)

# Define the base models to use in the ensemble
base_models = [
    ('lr', LinearRegression()),
    ('ridge', Ridge(random_state=seed)),
    ('lasso', Lasso(random_state=seed)),
    ('dt', DecisionTreeRegressor(random_state=seed)),
    ('rf', RandomForestRegressor(random_state=seed)),
    ('mlp', MLPRegressor(random_state=seed)),
    ('ada', AdaBoostRegressor(random_state=seed)),
    ('knr', KNeighborsRegressor()),
    ('gpr', GaussianProcessRegressor(random_state=seed))
]

# Define the meta-model to use in the ensemble
meta_model = LinearRegression()

# Initialize the Stacking Regressor
stacking_model = StackingRegressor(estimators=base_models, final_estimator=meta_model)

# Fit the Stacking Regressor to the training data
stacking_model.fit(X_train, y_train)

# Make predictions on the test set using the Stacking Regressor
y_pred = stacking_model.predict(X_test)

# Evaluate the performance of the Stacking Regressor using mean squared error, root mean squared error, and R-squared
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse:.4f}")
print(f"Root Mean Squared Error: {rmse:.4f}")
print(f"R-squared: {r2:.4f}")

Mean Squared Error: 0.0033
Root Mean Squared Error: 0.0577
R-squared: 0.8791


In [82]:
from sklearn.ensemble import StackingRegressor
from sklearn.linear_model import LinearRegression, Ridge, Lasso, ElasticNet
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor, AdaBoostRegressor
from sklearn.svm import SVR, LinearSVR
from sklearn.metrics import mean_squared_error, r2_score
import utlis
from sklearn.neural_network import MLPRegressor
from sklearn.neighbors import KNeighborsRegressor
from sklearn.gaussian_process import GaussianProcessRegressor
import xgboost as xgb

# Get data
X_train, y_train, X_test, y_test = utlis.split_data(df_reduced=df_reduced, split_rate=0.8)

nb_features = len(X_train.columns.tolist())
nb_targets = 1
layer_size = nb_features + nb_targets + 2

# Define the base models to use in the ensemble
base_models = [
    ('lr', LinearRegression()),
    ('ridge', Ridge(random_state=seed)),
    ('lasso', Lasso(random_state=seed)),
    ('dt_friedman_mse', DecisionTreeRegressor(criterion='friedman_mse', random_state=seed)),
    ('dt_absolute_error', DecisionTreeRegressor(criterion='absolute_error', random_state=seed)),
    ('dt_poisson', DecisionTreeRegressor(criterion='poisson', random_state=seed)),
    ('RFRS',RandomForestRegressor(criterion='squared_error', n_estimators=100, random_state = seed)), 
    ('RFRA', RandomForestRegressor(criterion='absolute_error', n_estimators=100, random_state = seed)), 
    ('RFRP', RandomForestRegressor(criterion='poisson', n_estimators=100, random_state = seed)), 
    ('mlp', MLPRegressor(random_state=seed)),
    ('ada', AdaBoostRegressor(random_state=seed)),
    ('knr', KNeighborsRegressor()),
    ('gpr', GaussianProcessRegressor(random_state=seed)),
    ('MLPR1', MLPRegressor(hidden_layer_sizes = (layer_size, ), max_iter=2000, early_stopping=True, random_state = seed)), 
    ('MLPR2', MLPRegressor(hidden_layer_sizes = (layer_size, layer_size,), max_iter=2000, early_stopping=True, random_state = seed)), 
    ('elastic_net', ElasticNet(alpha=0.01, l1_ratio=0.15, random_state=seed)),
    ('XGBR', xgb.XGBRegressor(objective ='reg:squarederror', colsample_bytree = 0.3, learning_rate = 0.1)),
    ('LSVR', LinearSVR(epsilon=11, C=37, fit_intercept=True, loss='squared_epsilon_insensitive', max_iter=30_000, random_state=seed)),
    ('SVR', SVR(kernel='poly', C=1e3, gamma=0.1, epsilon=0.1))
]

# Define the meta-model to use in the ensemble
meta_model = LinearRegression()

# Initialize the Stacking Regressor
stacking_model = StackingRegressor(estimators=base_models, final_estimator=meta_model, n_jobs=-1)

# Fit the Stacking Regressor to the training data
stacking_model.fit(X_train, y_train)

# Make predictions on the test set using the Stacking Regressor
y_pred = stacking_model.predict(X_test)

# Evaluate the performance of the Stacking Regressor using mean squared error, root mean squared error, and R-squared
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse:.4f}")
print(f"Root Mean Squared Error: {rmse:.4f}")
print(f"R-squared: {r2:.4f}")


In [ ]:
# save the model to disk
filename = 'stacking_model.sav'
pickle.dump(stacking_model, open(filename, 'wb'))

# 3.2 Model inspection and elimination <a id='model-inspection'></a>

In [ ]:
# Lets look at the importance of the different base models to the final stacking model
# Get the feature importance of the base models
# Get the feature importance of the base models from the coefficients of the meta-model
importance_weights = np.abs(stacking_model.final_estimator_.coef_)

# Normalize the importance weights
importance_weights = importance_weights / np.sum(importance_weights)

# Print the importance weights of the base models
for i, (model_name, _) in enumerate(base_models):
    print(f"{model_name}: {importance_weights[i]:.4f}")

            


lr: 0.2734
ridge: 0.3654
lasso: 0.0068
dt_friedman_mse: 0.0036
dt_absolute_error: 0.0017
dt_poisson: 0.0023
RFRS: 0.0328
RFRA: 0.0443
RFRP: 0.0030
mlp: 0.0127
ada: 0.0000
knr: 0.0056
gpr: 0.1772
MLPR1: 0.0120
MLPR2: 0.0004
elastic_net: 0.0225
XGBR: 0.0361


In [77]:
# eliminate all models under 0.05
base_models_f = [
    ('lr', LinearRegression()),
    ('ridge', Ridge(random_state=seed)),
    ('rf', RandomForestRegressor(random_state=seed)),
    ('gpr', GaussianProcessRegressor(random_state=seed)),
]

# Define the meta-model to use in the ensemble
meta_model = LinearRegression()

# Initialize the Stacking Regressor
stacking_model = StackingRegressor(estimators=base_models_f, final_estimator=meta_model)

# Fit the Stacking Regressor to the training data
stacking_model.fit(X_train, y_train)

# Make predictions on the test set using the Stacking Regressor
y_pred = stacking_model.predict(X_test)

# Evaluate the performance of the Stacking Regressor using mean squared error, root mean squared error, and R-squared
mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

print(f"Mean Squared Error: {mse:.4f}")
print(f"Root Mean Squared Error: {rmse:.4f}")
print(f"R-squared: {r2:.4f}")



Mean Squared Error: 0.0034
Root Mean Squared Error: 0.0586
R-squared: 0.8756


## 3.3 Final results <a id='final-results'></a>

In [1]:
!pip freeze > requirements.txt

Verwendung: java [Optionen] <mainclass> [args...]
           (zur Ausführung einer Klasse)
   oder  java [Optionen] -jar <jarfile> [args...]
           (zur Ausführung einer JAR-Datei)
   oder  java [Optionen] -m <module>[/<mainclass>] [args...]
       java [Optionen] --module <module>[/<mainclass>] [args...]
            (zur Ausführung der Hauptklasse in einem Modul)

 Argumente, die auf die Hauptklasse folgen, -jar <jarfile>, -m oder --module
 <module>/<mainclass> werden als Argumente für die Hauptklasse übergeben.

 wobei "Optionen" Folgendes umfasst:

    -cp <Klassensuchpfad mit Verzeichnissen und ZIP-/JAR-Dateien>
    -classpath <Klassensuchpfad mit Verzeichnissen und ZIP-/JAR-Dateien>
    --class-path <Klassensuchpfad mit Verzeichnissen und ZIP-/JAR-Dateien>
                  Eine durch : getrennte Liste mit Verzeichnissen, JAR-Archiven
                  und ZIP-Archiven, in denen nach Klassendateien gesucht wird.
    -p <Modulpfad>
    --module-path <Modulpfad>...
             